hypervisor/src/kvm.rs

```diff
@@ -0,0 +1,1269 @@
+        self.fd
+            .set_xcrs(&xcrs)
+            .map_err(|e| cpu::HypervisorCpuError::SetXcsr(e.into()))
+    }
+    /// Triggers the running of the current virtual CPU returning an exit reason.
+    ///
+    /// See documentation for `KVM_RUN`.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # extern crate hypervisor;
+    /// # use hypervisor::{KvmHypervisor, VcpuExit};
+    /// # use std::io::Write;
+    /// # use std::ptr::null_mut;
+    /// # use std::slice;
+    /// # use kvm_bindings::{kvm_userspace_memory_region, KVM_MEM_LOG_DIRTY_PAGES};
+    /// let hv = KvmHypervisor::new().unwrap();
+    /// let vm = hv.create_vm().unwrap();
+    /// // This is a dummy example for running on x86 based on https://lwn.net/Articles/658511/.
+    /// #[cfg(target_arch = "x86_64")] {
+    ///     let mem_size = 0x4000;
+    ///     let guest_addr: u64 = 0x1000;
+    ///     let load_addr: *mut u8 = unsafe {
+    ///         libc::mmap(
+    ///             null_mut(),
+    ///             mem_size,
+    ///             libc::PROT_READ | libc::PROT_WRITE,
+    ///             libc::MAP_ANONYMOUS | libc::MAP_SHARED | libc::MAP_NORESERVE,
+    ///             -1,
+    ///             0,
+    ///         ) as *mut u8
+    ///     };
+    ///
+    ///     let mem_region = kvm_userspace_memory_region {
+    ///         slot: 0,
+    ///         guest_phys_addr: guest_addr,
+    ///         memory_size: mem_size as u64,
+    ///         userspace_addr: load_addr as u64,
+    ///         flags: 0,
+    ///     };
+    ///     unsafe { vm.set_user_memory_region(mem_region).unwrap() };
+    ///
+    ///     // Dummy x86 code that just calls halt.
+    ///     let x86_code = [
+    ///             0xf4,             /* hlt */
+    ///     ];
+    ///
+    ///     // Write the code in the guest memory. This will generate a dirty page.
+    ///     unsafe {
+    ///         let mut slice = slice::from_raw_parts_mut(load_addr, mem_size);
+    ///         slice.write(&x86_code).unwrap();
+    ///     }
+    ///
+    ///     let vcpu_fd = vm.create_vcpu(0).unwrap();
+    ///
+    ///     let mut vcpu_sregs = vcpu_fd.get_sregs().unwrap();
+    ///     vcpu_sregs.cs.base = 0;
+    ///     vcpu_sregs.cs.selector = 0;
+    ///     vcpu_fd.set_sregs(&vcpu_sregs).unwrap();
+    ///
+    ///     let mut vcpu_regs = vcpu_fd.get_regs().unwrap();
+    ///     // Set the Instruction Pointer to the guest address where we loaded the code.
+    ///     vcpu_regs.rip = guest_addr;
+    ///     vcpu_regs.rax = 2;
+    ///     vcpu_regs.rbx = 3;
+    ///     vcpu_regs.rflags = 2;
+    ///     vcpu_fd.set_regs(&vcpu_regs).unwrap();
+    ///
+    ///     loop {
+    ///         match vcpu_fd.run().expect("run failed") {
+    ///             VcpuExit::Hlt => {
+    ///                 break;
+    ///             }
+    ///             exit_reason => panic!("unexpected exit reason: {:?}", exit_reason),
+    ///         }
+    ///     }
+    /// }
```

Comment:
I'm wondering if all these examples/documentation are needed since they've been ported from the `kvm-ioctls` crate. That feels like a lot of duplicates.

Reply:
Right. But the example has some changes reflecting the current hypervisor implementations.

Reply:
Right, but could we keep it shorter since the core content is already part of `kvm-ioctls`?

Reply:
Will do . thanks